src/v/cloud_storage/remote.cc

```diff
@@ -40,10 +43,149 @@ struct key_and_node {
+        break;
+    }
+    return o;
+}
+
+struct failure_injector::failure {
+    injected_failure_type _type{injected_failure_type::none};
+    ss::lowres_clock::duration _timeout{};
+
+    ss::future<std::optional<result<
+      http::client::response_stream_ref,
+      cloud_storage_clients::error_outcome>>>
+    generate_get_failure(ss::abort_source& as, retry_chain_logger& ctxlog) {
+        auto res = do_generate_response(true);
+        vlog(
+          ctxlog.warn,
+          "GET Error injected, type {}, timeout {}ns",
+          _type,
+          std::chrono::duration_cast<std::chrono::nanoseconds>(_timeout)
+            .count());
+        co_await ss::sleep_abortable(_timeout, as);
+        co_return res;
+    }
+
+    ss::future<std::optional<result<
+      cloud_storage_clients::client::no_response,
+      cloud_storage_clients::error_outcome>>>
+    generate_put_failure(ss::abort_source& as, retry_chain_logger& ctxlog) {
+        auto res = do_generate_response(false);
+        vlog(
+          ctxlog.warn,
+          "PUT Error injected, type {}, timeout {}ns",
+          _type,
+          std::chrono::duration_cast<std::chrono::nanoseconds>(_timeout)
+            .count());
+        co_await ss::sleep_abortable(_timeout, as);
+        co_return res;
+    }
+
+    std::optional<cloud_storage_clients::error_outcome>
+    do_generate_response(bool is_get_request) {
+        std::optional<cloud_storage_clients::error_outcome> res;
+        switch (_type) {
+        case injected_failure_type::failure:
+            res = cloud_storage_clients::error_outcome::fail;
+            break;
+        case injected_failure_type::no_such_key:
+            res = cloud_storage_clients::error_outcome::key_not_found;
+            if (is_get_request) {
+                break;
+            }
+        case injected_failure_type::throttle:
+            res = cloud_storage_clients::error_outcome::retry_slowdown;
+            break;
+        case injected_failure_type::none:
+            // In this case we only injecting sleep but not the actual
+            // failure.
+            res = std::nullopt;
+            break;
+        }
+        return res;
+    }
+};
+
+void failure_injector::load_from_string(const char* ctrl_string) {
+    std::stringstream str(ctrl_string);
+    std::string t;
+    if (std::getline(str, t, ':')) {
+        _freq = boost::lexical_cast<int32_t>(t);
+        while (std::getline(str, t, ':')) {
+            auto res = string_switch<injected_failure_type>(t)
+                         .match("failure", injected_failure_type::failure)
+                         .match("none", injected_failure_type::none)
+                         .match(
+                           "no_such_key", injected_failure_type::no_such_key)
+                         .match("throttle", injected_failure_type::throttle)
+                         .default_match(injected_failure_type::none);
+            _types.push_back(res);
+        }
+    }
+}
```

Comment:
Thinking a bit about how to generalize this, I wonder if it makes sense to add these to e.g. the node_config as a set of properties that encode these same errors.

Something like:

```
// Failure rates.
property<std::optional<float>> cloud_storage_remote_failure_rate;
property<std::optional<float>> cloud_storage_remote_no_such_key_rate;

// Defines some bounds with which to induce random latencies.
property<std::optional<int>> cloud_storage_remote_inject_latency_min_ms;
property<std::optional<int>> cloud_storage_remote_inject_latency_max_ms;
```

We should still be able to seed the RNG in a way that is somewhat deterministic, and I guess it wouldn't be all too different from what you have here, but reusing properties in this way may be more intuitive given we use them widely already today.